src/doc/man/svxreflector.conf.5
.TH SVXREFLECTOR.CONF 5 "JANUARI 2026" Linux "File Formats"
.
.SH NAME
.
svxreflector.conf \- Configuration file for the SvxReflector conference server audio reflector
.
.SH DESCRIPTION
.
The
.B svxreflector
server is used to interconnect multiple SvxLink nodes into one network.  This
man-page describe the configuration file format for it.
.P
SvxReflector look for configuration files in a number of places. First it tries
to find a user specific configuration file. It will look for a user specific
configuration file in:
.BR $HOME/.svxlink/svxreflector.conf .
If no user specific configuration file can be found, svxreflector will look
for the system wide configuration file
.BR /etc/svxlink/svxreflector.conf .
The
.B --config
command line option may also be used to specify an arbitrary configuration file.
.
.SH FILE FORMAT
.
The configuration file is in the famous INI-file format. A generic example of
how such a file might look like is shown below.

  [SECTION1]
  VALUE1=1
  VALUE2="TWO "
  VAULE3="Multi "
         "line"
  
  [SECTION2]
  VALUE1=2

This is a simple format that contain name=value pairs that belong to a section.
In written text, a specific configuration variable can be referred to as
SECTION1/VALUE2 meaning "configuration variable VALUE2 in section SECTION1".
.P
The same variable name can exist in two different sections. For example VALUE1
in section SECTION1 have the value 1 and VALUE1 in section SECTION2 have the
value 2. Values containing spaces at the beginning or end of the line must be
surrounded by citation characters (see SECTION1/VALUE2). Likewise with a multi
line value (see SECTION1/VALUE3).
.
.SH CONFIGURATION VARIABLES
.
Here is the description of all configuration variables that svxreflector
understands. The configuration variables are described section for section.
.
.SS GLOBAL
.
The GLOBAL section contains application global configuration data.
.TP
.B CFG_DIR
Specify the path to a directory that contain additional configuration files.
If a relative path is specified, the path will be relative to the directory
where the main configuration file is at. All files in the specified directory
will be read as additional configuration. Filenames starting with a dot are
ignored.
.TP
.B TIMESTAMP_FORMAT
This variable specifies the format of the timestamp that is written in front of
each row in the log file. The format string is in the same format as specified
in the
.BR strftime (3)
manual page. The default is "%c" which is described as: "the preferred date and
time representation for the current locale". The environment variables LC_TIME,
LC_ALL and LANG will affect how this time format will look. For example, setting
LC_TIME="sv_SE.UTF8" will give you Swedish timestamp representation. Other
examples of format specifiers are:
.RS
.IP \(bu 4
.BR %d " - The day of the month as a decimal number (range 01 to 31)"
.IP \(bu 4
.BR %b " - The abbreviated month name according to the current locale"
.IP \(bu 4
.BR %Y " - The year as a decimal number including the century"
.IP \(bu 4
.BR %H " - The hour as a decimal number using a 24-hour clock (range 00 to 23)"
.IP \(bu 4
.BR %M " - The minute as a decimal number (range 00 to 59)"
.IP \(bu 4
.BR %S " - The second as a decimal number (range 00 to 61)"
.P
So, TIMESTAMP_FORMAT="%d %b %Y %H:%M:%S" would give a timestamp looking something like:
"29 Nov 2005 22:31:59".
.RE
.TP
.B LISTEN_PORT
The TCP and UDP port number to use for network communications. The default is
5300. Make sure to open this port for incoming traffic to the server on both
TCP and UDP. Clients do not have to open any ports in their firewalls.
.TP
.B SQL_TIMEOUT
Use this configuration variable to set a time in seconds after which a clients
audio is blocked if he has been talking for too long. The default is 0
(disabled).
.TP
.B SQL_TIMEOUT_BLOCKTIME
If a client has been talking for longer than the time specified in the
.B SQL_TIMEOUT
configuration variable he will not be able to talk again until the time in this
configuration variable have elapsed. If not specified, the default is one
second.
.TP
.B CODECS
A comma separated list of allowed codecs. For the moment only one codec can be
specified. Choose from the following codecs: OPUS, SPEEX, GSM, S16
(uncompressed signed 16 bit), RAW (uncompressed 32 bit floats). The default is
OPUS and you should have a very good reason for changing this since that codec
provide both low bandwidth (~20kbps by default) and very good audio quality.
.TP
.B TG_FOR_V1_CLIENTS
Set which talk group to place protocol version 1 clients in. Without this
configuration version 1 clients will not be able to use the reflector since
they cannot select a talk group by themselves. Talk groups should normally be
configured to start with the ITU-T E.212 Mobile Country Code (MCC) for your
country, e.g. 240 for Sweden.
.TP
.B RANDOM_QSY_RANGE
Specify in which talk group range the reflector server should select random
talk groups used when using the QSY functionality. The range is specified using
two colon separated values. The first one is the lower limit and the second
value is the number of talk groups in the range. As an example, if set to
1239900:100 QSY talk groups will be selected from the range 1239900 to 1239999.
The recommended lower range to use is <MCC>9900 where MCC is the ITU-T E.212
Mobile Country Code for your country, e.g. 240 for Sweden.
.TP
.B HTTP_SRV_PORT
Set which port to use for the HTTP server. The HTTP server can be used to fetch
reflector status. No port is set by default. Don't expose this port to the
public Internet. There are multiple reasons for that. The built in web server
is very simple and could be incompatible with some clients. It has not been
audited for security problems so there may be security issues. There is also
the risk of some client overwhelming the reflector with requests causing
disturbances in the reflector operation.

Example: HTTP_SRV_PORT=8080
.TP
.B AUTH_KEY
Set a key used by nodes that are not configured in [USERS] section without
defining an USER=password combination. This can be used for reflectors that
should be open for everyone and/or a big number of clients to keep the 
configuration effort as low as possible.
.TP
.B COMMAND_PTY
Configure a path for a pseudo tty device to send runtime commands to the
svxreflector. The device may be defined as COMMAND_PTY=/dev/shm/reflector_ctrl.
See the "COMMAND PTY" section below for more information on how to use the
command PTY.
.TP
.B ACCEPT_CALLSIGN
A regular expression for verifying that a received login callsign has the
correct format. The default is to roughly match standard ham radio callsigns.
Example: ACCEPT_CALLSIGN="S[A-M]\\d[A-Z]{1,3}(?:-[A-Z0-9]{1,3})?"
.TP
.B REJECT_CALLSIGN
A regular expression for rejecting login callsigns. This can be used to block
some specific callsigns or even a whole range of callsigns. No callsigns are
rejected by default.
Example: REJECT_CALLSIGN="SM0XYZ|SM1ABC|SM2.*"
.TP
.B ACCEPT_CERT_EMAIL
A regular expression for verifying that email addresses in a received
certificate signing requests are valid. If unset, which is the default, all
email addresses are accepted. If an invalid email address is received from a
node it will be disconnected and an error message will be sent to the node. The
following example should accept most email addresses at the same time as
detecting most invalid ones.
Example: ACCEPT_CERT_EMAIL=""\\w+(?:[-._+]\\w+)*@\\w+(?:\\.\\w+)*"
.TP
.B CERT_PKI_DIR
The path to the directory containing PKI (Public Key Infrastructure) files. If
a relative path is given, the value of the build time variable
SVX_LOCAL_STATE_DIR (e.g. /var/lib/svxlink/) will be prepended.

Default: CERT_PKI_DIR=pki/
.TP
.B CERT_CA_KEYS_DIR
The path to the directory containing private key files. If
a relative path is given, the value of the CERT_PKI_DIR variable will be
prepended.

Default: CERT_CA_KEYS_DIR=private/
.TP
.B CERT_CA_PENDING_CSRS_DIR
The path to the directory containing unsigned CSR files. If a relative path is
given, the value of the CERT_PKI_DIR variable will be prepended.

Default: CERT_CA_PENDING_CSRS_DIR=pending_csrs/
.TP
.B CERT_CA_CSRS_DIR
The path to the directory containing signed CSR files. If a relative path is
given, the value of the CERT_PKI_DIR variable will be prepended.

Default: CERT_CA_CSRS_DIR=csrs/
.TP
.B CERT_CA_CERTS_DIR
The path to the directory containing signed certificate files. If a relative
path is given, the value of the CERT_PKI_DIR variable will be prepended.

Default: CERT_CA_CERTS_DIR=certs/
.TP
.B CERT_CA_HOOK
Set to a path for an external application that is run on CA events like when a
CSR is received or when a certificate is signed. The following environment
variables will be set up before running the external application.
.RS
.IP \(bu 4
.BR "CA_OP" ":"
Set to one of the strings PENDING_CSR_CREATE, PENDING_CSR_UPDATE or
CSR_SIGNED, which should be self explanatory.
.IP \(bu 4
.BR "CA_CSR_PEM" ":"
CSR data in PEM format. This is not set for operation CSR_SIGNED.
.IP \(bu 4
.BR "CA_CRT_PEM" ":"
Certificate data in PEM format. This is only set for operation CSR_SIGNED.
.RE
.
.SS ROOT_CA, ISSUING_CA and SERVER_CERT sections
.
These configuration sections is used to customize the contents of the different
certificates that is generated by the reflector server.
.TP
.B KEYFILE
The path to the certificate private key file. If a relative path is given, the
value of the CERT_CA_KEYS_DIR variable will be prepended.
.TP
.B CSRFILE
The path to the certificate signing request file. If a relative path is given,
the value of the CERT_CA_CSRS_DIR variable will be prepended.
.TP
.B CRTFILE
The path to the certificate file. If a relative path is given, the value of the
CERT_CA_CERTS_DIR variable will be prepended.
.TP
.B COMMON_NAME
The common name (CN) used in the subject of the certificate.
.TP
.B ORG_UNIT
The organizational unit (OU) used in the subject of the certificate.

Example: ORG_UNIT="VHF/UHF-sektionen"
.TP
.B ORG
The organization name (O) used in the subject of the certificate.

Example: ORG=SSA
.TP
.B LOCALITY
The locality name (L) used in the subject of the certificate.

Example: LOCALITY=Boden
.TP
.B STATE
The name of the state or province (ST) used in the subject of the certificate.

Example: STATE=Norrbotten
.TP
.B COUNTRY
The ISO country code (C) used in the subject of the certificate.

Example: COUNTRY=SE
.TP
.B SUBJECT_ALT_NAME
A comma separated list of subject alternative names.

Example: SUBJECT_ALT_NAME=DNS:public-hostname.example.org,IP:172.17.1.42
.TP
.B EMAIL_ADDRESS
The email address that can be used to get in contact with the owner of the
certificate.

Example: EMAIL_ADDRESS=sysop@svxlink.example.org
.
.SS USERS and PASSWORDS sections
.
Previously, all users had to be specified in the USERS section. That is not
needed anymore for nodes that use a SvxLink version > 1.8.0. Usernames are
totally arbitrary but typically the callsign is used. The purpose is to map
each username to a password. However, the password is not given directly but
instead a reference to the PASSWORDS section is given. It's done like this so
that multiple nodes can share a single password.
.P
An example of how the USERS and PASSWORDS sections may look is given below.

  [USERS]
  SM0ABC-1=MyNodes
  SM0ABC-2=MyNodes
  SM1XYZ=SM1XYZ
  
  [PASSWORDS]
  MyNodes="A very strong password!"
  SM1XYZ="Another very good password?"

This will set up
.BR SM0ABC-1 " and " SM0ABC-2
to use the same password specified by
.B MyNodes
in the PASSWORDS section. User
.BR SM1XYZ " have his own password."
.
.SS Talkgroup Configuration Sections
.
It is possible to set configuration parameters that are only applied to one
talkgroup. Those parameters are placed in a configuration section named
"TG#<talkgroup id>". Example:

  [TG#9999]
  AUTO_QSY_AFTER=300
  ALLOW=S[A-M]\\\\d.*|LA8PV
  ALLOW_MONITOR=S[A-M]3.*
  SHOW_ACTIVITY=0

The following configuration variables are valid in a talkgroup configuration
section.
.TP
.B AUTO_QSY_AFTER
Set this to the number of seconds after which an automatic QSY to a random
talkgroup is requested. This is typically used to keep call channels and wide
area channels clear so that stations does not dwell there for too long.
Auto QSY is only triggered directly aftar a talker stop event.
The default is that auto QSY is disabled (AUTO_QSY_AFTER=0).
.TP
.B ALLOW
A regular expression that must match the whole callsign of the nodes that try
to select this talkgroup.  The regular expression standard used is the same as
for ECMAScript with the exception that the backslash special patterns (e.g.
\\s, \\d etc) must be expressed with double backslash notation (e.g. \\\\s,
\\\\d etc).

An example that match most Swedish callsigns and a specific Norwegian one is
"S[A-M]\\\\d.*|LA8PV". That expression will for example match LA8PV, SM0SVX,
SK3W, SA7ABC etc.
.TP
.B ALLOW_MONTOR
A regular expression that must match the whole callsign of the nodes that try
to monitor this talkgroup.  The regular expression standard used is the same as
for ECMAScript with the exception that the backslash special patterns (e.g.
\\s, \\d etc) must be expressed with double backslash notation (e.g. \\\\s,
\\\\d etc).

An example that match most Swedish callsigns in district 3 is
"S[A-M]3.*". That expression will for example match SA3ABC, SM3XYZ, SK3GK,
SH3YZX etc.
.TP
.B SHOW_ACTIVITY
If set to 0, do not indicate in the http status message when the talkgroup is
in use by a node. Default is 1 = show activity.
.
.SH COMMAND PTY
.
If a command PTY has been set up using the COMMAND_PTY configuration variable
it is possible to use the commands described below to control some functions in
the reflector server. The commands can be sent to the PTY by using the
following shell command:

.B "  echo THE COMMAND >/dev/shm/reflector_ctrl"

Valid commands are:
.TP
.B CFG <section> <tag> <value>
Read or set configuration. Configuration variables that is set is not persisted
so the new setting will be lost on next restart. If the command is used without
arguments, the full configuration will be listed. If only a section is given,
that section will be listed. If both section and tag is given, only that
specific configuration variable will be listed. If in addition a value is
given, that specific configuration variable will be set to that value.
.TP
.B NODE BLOCK <callsign> <blocktime seconds>
Temporarily block a node from being a talker for the given number of seconds.
The node will still be able to choose the talkgroup and listen to the audio but
will not be able to participate in the QSO. This command is good to use when a
node is disturbing a QSO for whatever reason. The blocktime till be reset to
the given value each time the node try to send audio to the reflector. Setting
a blocktime of 0 will remove the block. The block is also removed if the node
disconnects.
.TP
.B CA LS|LSC|LSP|SIGN <callsign>|RM <callsign>
The
.B CA
command is used to manipulate the Certificate Authority part of the reflector
server.
.br
.B CA LS
is used to list all certificates and pending CSRs (Certificate Signing Requests).
.br
.BR "CA LSC" " and " "CA LSP"
are used to list only the certificates or pending CSRs respectively.
.br
.B CA SIGN
is used to sign a pending CSR for the given callsign. Note that the callsign is
case-sensitive.
.br
.B CA RM
Remove all certificate files associated with the given callsign. Note that it
is NOT sufficient to remove the files to stop the given callsign from logging
in. If the node already has a valid certificate, it can be used to log in. To
stop a node from logging in, use the REJECT_CALLSIGN configuration variable.
.
.SH FILES
.
.TP
.I /etc/svxlink/svxreflector.conf
The system wide configuration file.
.TP
.I ~/.svxlink/svxreflector.conf
Per user configuration file.
.TP
.I /etc/svxlink/svxreflector.d/*
Additional configuration files. This directory is setup by the CFG_DIR
configuration variable.
.
.SH AUTHOR
.
Tobias Blomberg (SM0SVX) <sm0svx at svxlink dot org>
.
.SH REPORTING BUGS
.
Bugs should be reported using the issue tracker at
https://github.com/sm0svx/svxlink.

Questions about SvxLink should not be asked using the issue tracker. Instead
use the group set up for this purpose at groups.io:
https://groups.io/g/svxlink
.
.SH "SEE ALSO"
.
.BR svxreflector (1),
.BR svxlink (1),
.BR svxlink.conf (5),


src/svxlink/reflector/svxreflector.cpp
/**
@file	 svxreflector.cpp
@brief   Main source file for the SvxReflector application
@author  Tobias Blomberg / SM0SVX
@date	 2017-02-11

The SvxReflector application is a central hub used to connect multiple SvxLink
nodes together.

\verbatim
SvxReflector - An audio reflector for connecting SvxLink Servers
Copyright (C) 2003-2025 Tobias Blomberg / SM0SVX

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
\endverbatim
*/



/****************************************************************************
 *
 * System Includes
 *
 ****************************************************************************/

#include <locale.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <grp.h>
#include <pwd.h>
#include <dirent.h>
#include <termios.h>
#include <errno.h>

#include <popt.h>
#include <sigc++/sigc++.h>

#include <cstdio>
#include <cstring>
#include <iostream>
#include <iomanip>


/****************************************************************************
 *
 * Project Includes
 *
 ****************************************************************************/

#include <AsyncCppApplication.h>
#include <AsyncFdWatch.h>
#include <AsyncConfig.h>
#include <config.h>
#include <LogWriter.h>


/****************************************************************************
 *
 * Local Includes
 *
 ****************************************************************************/

#include "version/SVXREFLECTOR.h"
#include "Reflector.h"


/****************************************************************************
 *
 * Namespaces to use
 *
 ****************************************************************************/

using namespace std;
using namespace Async;


/****************************************************************************
 *
 * Defines & typedefs
 *
 ****************************************************************************/

#define PROGRAM_NAME "SvxReflector"


/****************************************************************************
 *
 * Local class definitions
 *
 ****************************************************************************/



/****************************************************************************
 *
 * Prototypes
 *
 ****************************************************************************/

static void parse_arguments(int argc, const char **argv);
static void stdinHandler(FdWatch *w);
static void sighup_handler(int signal);
static void sigterm_handler(int signal);
static void handle_unix_signal(int signum);


/****************************************************************************
 *
 * Exported Global Variables
 *
 ****************************************************************************/



/****************************************************************************
 *
 * Local Global Variables
 *
 ****************************************************************************/

namespace {
  char*                 pidfile_name = nullptr;
  char*                 logfile_name = nullptr;
  char*                 runasuser = nullptr;
  char*                 config = nullptr;
  int                   daemonize = 0;
  int                   quiet = 0;
  FdWatch*              stdin_watch = 0;
  LogWriter             logwriter;
};


/****************************************************************************
 *
 * MAIN
 *
 ****************************************************************************/

/*
 *----------------------------------------------------------------------------
 * Function:  main
 * Purpose:   Start everything...
 * Input:     argc  - The number of arguments passed to this program
 *    	      	      (including the program name).
 *    	      argv  - The arguments passed to this program. argv[0] is the
 *    	      	      program name.
 * Output:    Return 0 on success, else non-zero.
 * Author:    Tobias Blomberg, SM0SVX
 * Created:   2017-02-11
 * Remarks:   
 * Bugs:      
 *----------------------------------------------------------------------------
 */
int main(int argc, const char *argv[])
{
  setlocale(LC_ALL, "");

  CppApplication app;
  app.catchUnixSignal(SIGHUP);
  app.catchUnixSignal(SIGINT);
  app.catchUnixSignal(SIGTERM);
  app.unixSignalCaught.connect(sigc::ptr_fun(&handle_unix_signal));

  parse_arguments(argc, const_cast<const char **>(argv));

  if (daemonize && (daemon(1, 0) == -1))
  {
    perror("daemon");
    exit(1);
  }

  if (quiet || (logfile_name != 0))
  {
    int devnull = open("/dev/null", O_RDWR);
    if (devnull == -1)
    {
      perror("open(/dev/null)");
      exit(1);
    }

    if (quiet)
    {
        /* Redirect stdout to /dev/null */
      dup2(devnull, STDOUT_FILENO);
    }

    if (logfile_name != 0)
    {
      logwriter.setDestinationName(logfile_name);
      if (!quiet)
      {
        logwriter.redirectStdout();
      }
      logwriter.redirectStderr();
      logwriter.start();

        /* Redirect stdin to /dev/null */
      if (dup2(devnull, STDIN_FILENO) == -1)
      {
        perror("dup2(stdin)");
        exit(1);
      }
    }
    close(devnull);
  }

  if (pidfile_name != NULL)
  {
    FILE *pidfile = fopen(pidfile_name, "w");
    if (pidfile == 0)
    {
      char err[256];
      sprintf(err, "fopen(\"%s\")", pidfile_name);
      perror(err);
      fflush(stderr);
      exit(1);
    }
    fprintf(pidfile, "%d\n", getpid());
    fclose(pidfile);
  }

  const char *home_dir = 0;
  if (runasuser != NULL)
  {
      // Setup supplementary group IDs
    if (initgroups(runasuser, getgid()))
    {
      perror("initgroups");
      exit(1);
    }

    struct passwd *passwd = getpwnam(runasuser);
    if (passwd == NULL)
    {
      perror("getpwnam");
      exit(1);
    }
    if (setgid(passwd->pw_gid) == -1)
    {
      perror("setgid");
      exit(1);
    }
    if (setuid(passwd->pw_uid) == -1)
    {
      perror("setuid");
      exit(1);
    }
    home_dir = passwd->pw_dir;
  }

  if (home_dir == 0)
  {
    home_dir = getenv("HOME");
  }
  if (home_dir == 0)
  {
    home_dir = ".";
  }

  Config cfg;
  string cfg_filename;
  if (config != NULL)
  {
    cfg_filename = string(config);
    if (!cfg.open(cfg_filename))
    {
      cerr << "*** ERROR: Could not open configuration file: "
           << config << endl;
      exit(1);
    }
  }
  else
  {
    cfg_filename = string(home_dir);
    cfg_filename += "/.svxlink/svxreflector.conf";
    if (!cfg.open(cfg_filename))
    {
      cfg_filename = SVX_SYSCONF_INSTALL_DIR "/svxreflector.conf";
      if (!cfg.open(cfg_filename))
      {
        cerr << "*** ERROR: Could not open configuration file";
        if (errno != 0)
        {
          cerr << " (" << strerror(errno) << ")";
        }
        cerr << ".\n";
        cerr << "Tried the following paths:\n"
             << "\t" << home_dir << "/.svxlink/svxreflector.conf\n"
             << "\t" SVX_SYSCONF_INSTALL_DIR "/svxreflector.conf\n"
             << "Possible reasons for failure are: None of the files exist,\n"
             << "you do not have permission to read the file or there was a\n"
             << "syntax error in the file.\n";
        exit(1);
      }
    }
  }
  string main_cfg_filename(cfg_filename);

  std::string main_cfg_dir = ".";
  auto slash_pos = main_cfg_filename.rfind('/');
  if (slash_pos != std::string::npos)
  {
    main_cfg_dir = main_cfg_filename.substr(0, slash_pos);
  }

  string cfg_dir;
  if (cfg.getValue("GLOBAL", "CFG_DIR", cfg_dir))
  {
    if (cfg_dir[0] != '/')
    {
      cfg_dir = main_cfg_dir + "/" + cfg_dir;
    }

    DIR *dir = opendir(cfg_dir.c_str());
    if (dir == NULL)
    {
      cerr << "*** ERROR: Could not read from directory spcified by "
           << "configuration variable GLOBAL/CFG_DIR=" << cfg_dir << endl;
      exit(1);
    }

    struct dirent *dirent;
    while ((dirent = readdir(dir)) != NULL)
    {
      char *dot = strrchr(dirent->d_name, '.');
      if ((dot == NULL) || (dirent->d_name[0] == '.') ||
          (strcmp(dot, ".conf") != 0))
      {
        continue;
      }
      cfg_filename = cfg_dir + "/" + dirent->d_name;
      if (!cfg.open(cfg_filename))
       {
	 cerr << "*** ERROR: Could not open configuration file: "
	      << cfg_filename << endl;
	 exit(1);
       }
    }

    if (closedir(dir) == -1)
    {
      cerr << "*** ERROR: Error closing directory specified by"
           << "configuration variable GLOBAL/CFG_DIR=" << cfg_dir << endl;
      exit(1);
    }
  }

  std::string tstamp_format = "%c";
  cfg.getValue("GLOBAL", "TIMESTAMP_FORMAT", tstamp_format);
  logwriter.setTimestampFormat(tstamp_format);

  cout << PROGRAM_NAME " v" SVXREFLECTOR_VERSION
          " Copyright (C) 2003-2025 Tobias Blomberg / SM0SVX\n\n";
  cout << PROGRAM_NAME " comes with ABSOLUTELY NO WARRANTY. "
          "This is free software, and you are\n";
  cout << "welcome to redistribute it in accordance with the "
          "terms and conditions in the\n";
  cout << "GNU GPL (General Public License) version 2 or later.\n";
  cout << "This is a special SvxReflector version for Tetra users\n";
  cout << "modified by Adi / DL1HRC\n";

  cout << "\nUsing configuration file: " << main_cfg_filename << endl;

  struct termios org_termios = {0};
  if (logfile_name == 0)
  {
    struct termios termios;
    tcgetattr(STDIN_FILENO, &org_termios);
    termios = org_termios;
    termios.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &termios);

    stdin_watch = new FdWatch(STDIN_FILENO, FdWatch::FD_WATCH_RD);
    stdin_watch->activity.connect(sigc::ptr_fun(&stdinHandler));
  }

  Reflector ref;
  if (ref.initialize(cfg))
  {
    std::cout << "NOTICE: Initialization done. Starting main application."
              << std::endl;
    app.exec();
  }
  else
  {
    cerr << ":-(" << endl;
  }

  if (stdin_watch != 0)
  {
    delete stdin_watch;
    tcsetattr(STDIN_FILENO, TCSANOW, &org_termios);
  }

  return 0;
} /* main */


/****************************************************************************
 *
 * Functions
 *
 ****************************************************************************/

/*
 *----------------------------------------------------------------------------
 * Function:  parse_arguments
 * Purpose:   Parse the command line arguments.
 * Input:     argc  - Number of arguments in the command line
 *    	      argv  - Array of strings with the arguments
 * Output:    Returns 0 if all is ok, otherwise -1.
 * Author:    Tobias Blomberg, SM0SVX
 * Created:   2000-06-13
 * Remarks:   
 * Bugs:      
 *----------------------------------------------------------------------------
 */
static void parse_arguments(int argc, const char **argv)
{
  int print_version = 0;

  poptContext optCon;
  const struct poptOption optionsTable[] =
  {
    POPT_AUTOHELP
    {"pidfile", 0, POPT_ARG_STRING, &pidfile_name, 0,
            "Specify the name of the pidfile to use", "<filename>"},
    {"logfile", 0, POPT_ARG_STRING, &logfile_name, 0,
            "Specify the logfile to use (stdout and stderr)", "<filename>"},
    {"runasuser", 0, POPT_ARG_STRING, &runasuser, 0,
            "Specify the user to run SvxLink as", "<username>"},
    {"config", 0, POPT_ARG_STRING, &config, 0,
	    "Specify the configuration file to use", "<filename>"},
    /*
    {"int_arg", 'i', POPT_ARG_INT, &int_arg, 0,
	    "Description of int argument", "<an int>"},
    */
    {"daemon", 0, POPT_ARG_NONE, &daemonize, 0,
	    "Start " PROGRAM_NAME " as a daemon", NULL},
    {"version", 0, POPT_ARG_NONE, &print_version, 0,
	    "Print the application version string", NULL},
    {NULL, 0, 0, NULL, 0}
  };
  int err;
  //const char *arg = NULL;
  //int argcnt = 0;

  optCon = poptGetContext(PROGRAM_NAME, argc, argv, optionsTable, 0);
  poptReadDefaultConfig(optCon, 0);

  err = poptGetNextOpt(optCon);
  if (err != -1)
  {
    fprintf(stderr, "\t%s: %s\n",
	    poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
	    poptStrerror(err));
    exit(1);
  }

  /*
  printf("string_arg  = %s\n", string_arg);
  printf("int_arg     = %d\n", int_arg);
  printf("bool_arg    = %d\n", bool_arg);
  */

    /* Parse arguments that do not begin with '-' (leftovers) */
  /*
  arg = poptGetArg(optCon);
  while (arg != NULL)
  {
    printf("arg %2d      = %s\n", ++argcnt, arg);
    arg = poptGetArg(optCon);
  }
  */

  poptFreeContext(optCon);

  if (print_version)
  {
    std::cout << SVXREFLECTOR_VERSION << std::endl;
    exit(0);
  }
} /* parse_arguments */


static void stdinHandler(FdWatch *w)
{
  char buf[1];
  int cnt = ::read(STDIN_FILENO, buf, 1);
  if (cnt == -1)
  {
    fprintf(stderr, "*** ERROR: Reading from stdin failed\n");
    Application::app().quit();
    return;
  }
  else if (cnt == 0)
  {
      /* Stdin file descriptor closed */
    delete stdin_watch;
    stdin_watch = 0;
    return;
  }

  switch (toupper(buf[0]))
  {
    case 'Q':
      Application::app().quit();
      break;

    case '\n':
      putchar('\n');
      break;

    default:
      break;
  }
} /* stdinHandler */


static void sighup_handler(int signal)
{
  if (logfile_name == 0)
  {
    cout << "Ignoring SIGHUP\n";
    return;
  }
  std::cout << "SIGHUP received" << std::endl;
  logwriter.reopenLogfile();
} /* sighup_handler */


static void sigterm_handler(int signal)
{
  const char *signame = 0;
  switch (signal)
  {
    case SIGTERM:
      signame = "SIGTERM";
      break;
    case SIGINT:
      signame = "SIGINT";
      break;
    default:
      signame = "Unknown signal";
      break;
  }

  std::cout << "\nNOTICE: " << signame
            << " received. Shutting down application..."
            << std::endl;
  Application::app().quit();
} /* sigterm_handler */


static void handle_unix_signal(int signum)
{
  switch (signum)
  {
    case SIGHUP:
      sighup_handler(signum);
      break;
    case SIGINT:
    case SIGTERM:
      sigterm_handler(signum);
      break;
  }
} /* handle_unix_signal */


/*
 * This file has not been truncated
 */

